### PR TITLE
ITE: drivers/serial: add console input expired

### PR DESCRIPTION
**Pull request by GTLin08**
If there are no console messages input at an interval of 15
seconds, the system will be able to enter suspend mode.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

### CODE REVIEW

File: drivers/serial/uart_ite_it8xxx2.c [hunk 1]
```diff
@@ -11,36 +11,48 @@
...
 
-	if (uart1_wui_isr_late) {
-		uart1_wui_isr_late();
-	}
+#ifdef CONFIG_UART_CONSOLE_INPUT_EXPIRED
+	k_timeout_t delay = K_MSEC(CONFIG_UART_CONSOLE_INPUT_EXPIRED_TIMEOUT);
+
+	pm_constraint_set(PM_STATE_STANDBY);
+	k_work_reschedule(&uart_data->rx_refresh_timeout_work, delay);
+#endif
```
Comment (brockus-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:55]: nit: Can you add a comment for readers as to why you are going standby and why it is only needed for UART1?

Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c:62]: Added comment. 
And added to judge whether uart is a console device to use console input expired.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 2]
```diff
@@ -28,19 +42,24 @@ enum uart_port_num {
...
 {
 	/* Disable interrupts on UART1 RX pin to avoid repeated interrupts. */
 	(void)gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
 					   GPIO_INT_DISABLE);
 
-	if (uart1_wui_isr_late) {
-		uart1_wui_isr_late();
-	}
+	/* Refresh console expired time if got UART Rx wake-up event */
+#ifdef CONFIG_UART_CONSOLE_INPUT_EXPIRED
```
Comment (brockus-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:53]: should this be "if (IS_ENABLED(CONFIG_UART_CONSOLE_INPUT_EXPIRED)) {"?

File: drivers/serial/uart_ite_it8xxx2.c [hunk 3]
```diff
@@ -50,9 +69,17 @@ void uart2_wui_isr(const struct device *gpio, struct gpio_callback *cb,
 	(void)gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
 					   GPIO_INT_DISABLE);
 
-	if (uart2_wui_isr_late) {
-		uart2_wui_isr_late();
-	}
+	/* Refresh console expired time if got UART Rx wake-up event */
+#ifdef CONFIG_UART_CONSOLE_INPUT_EXPIRED
```
Comment (brockus-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:73]: should this be "if (IS_ENABLED(CONFIG_UART_CONSOLE_INPUT_EXPIRED)) {"?

File: drivers/serial/uart_ite_it8xxx2.c [hunk 4]
```diff
@@ -84,40 +111,64 @@ static inline int uart_it8xxx2_pm_action(const struct device *dev,
...
-	} else if (config->port == UART2) {
-		static struct gpio_callback uart2_wui_cb;
-
-		gpio_init_callback(&uart2_wui_cb, uart2_wui_isr,
-				   BIT(config->gpio_wui.pin));
-
-		ret = gpio_add_callback(config->gpio_wui.port, &uart2_wui_cb);
-	}
+	if (config->uart_dev == uart_console_dev) {
+#ifdef CONFIG_UART_CONSOLE_INPUT_EXPIRED
```
Comment (brockus-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:140]: should this be "if (IS_ENABLED(CONFIG_UART_CONSOLE_INPUT_EXPIRED)) {"?

File: drivers/serial/uart_ite_it8xxx2.c (hunk 2 above)
Comment (keith-zephyr) [reply drivers/serial/uart_ite_it8xxx2.c:53]: It looks like the symbol `rx_refresh_timeout_work` is only available if `CONFIG_UART_CONSOLE_INPUT_EXPIRED`, so I think the #ifdef is required.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 5]
```diff
@@ -11,15 +11,29 @@
...
 #include <pm/device.h>
+#include <pm/pm.h>
+#include <pm/policy.h>
 #include <soc.h>
 
 #include <logging/log.h>
 LOG_MODULE_REGISTER(uart_ite_it8xxx2, LOG_LEVEL_ERR);
 
+#if defined(CONFIG_PM_DEVICE) && defined(CONFIG_UART_CONSOLE_INPUT_EXPIRED)
+static struct uart_it8xxx2_data *uart_data;
```
Comment (keith-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:22]: Nit - rename to `uart_console_data`.

Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

**Review by jackrosenthal — APPROVED** [review]

File: drivers/serial/uart_ite_it8xxx2.c (hunk 2 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c:53]: Ack.

**Comment by GTLin08** [comment]
Hi @keith-zephyr @brockus-zephyr ,
Any comments will be much appreciated.

**Review by keith-zephyr — APPROVED** [review]

**Pull request merged**